logger/logger_test.go

```diff
@@ -37,90 +46,212 @@ func (writer *mockWriter) Read() (logMsg, error) {
+		assert.Equal(t, tc.output, output, fmt.Sprintf("%s: expected %s got %s", tc.desc, tc.output, output))
 	}
 }
 
 func TestError(t *testing.T) {
-	cases := map[string]struct {
+	cases := []struct {
+		desc   string
 		input  string
 		output logMsg
 	}{
-		"error log ordinary string": {"input_string", logMsg{log.Error.String(), "input_string"}},
-		"error log empty string":    {"", logMsg{log.Error.String(), ""}},
+		{
+			desc:   "error log ordinary string",
+			input:  "input_string",
+			output: logMsg{log.Error.String(), "input_string", ""},
+		},
+		{
+			desc:   "error log empty string",
+			input:  "",
+			output: logMsg{log.Error.String(), "", ""},
+		},
 	}
 
 	writer := mockWriter{}
-	logger, _ := log.New(&writer, log.Error.String())
-
-	for desc, tc := range cases {
+	logger, err := log.New(&writer, log.Error.String())
+	require.Nil(t, err)
+	for _, tc := range cases {
 		logger.Error(tc.input)
 		output, err := writer.Read()
-		assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", desc, err))
-		assert.Equal(t, tc.output, output, fmt.Sprintf("%s: expected %s got %s", desc, tc.output, output))
+		assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+		assert.Equal(t, tc.output, output, fmt.Sprintf("%s: expected %s got %s", tc.desc, tc.output, output))
+	}
+}
+
+func TestFatal(t *testing.T) {
+	// This is the actually Fatal call we test that will
+	// be executed in the subprocess spawned by the test.
+	if os.Getenv(testFlag) == testFlagVal {
+		logger, err := log.New(os.Stderr, log.Error.String())
+		require.Nil(t, err)
+		msg := os.Getenv(testMsg)
+		logger.Fatal(msg)
+		return
+	}
+
+	cases := []struct {
+		desc   string
+		input  string
+		output logMsg
+	}{
+		{
+			desc:   "error log ordinary string",
+			input:  "input_string",
+			output: logMsg{"", "", "input_string"},
+		},
+		{
+			desc:   "error log empty string",
+			input:  "",
+			output: logMsg{"", "", ""},
+		},
+	}
+	writer := mockWriter{}
+	for _, tc := range cases {
+		// This command will run this same test as a separate subprocess.
+		// It needs to be executed as a subprocess because we need to test os.Exit(1) call.
+		cmd := exec.Command(os.Args[0], "-test.run=TestFatal")
+		// This flag is used to prevent an infinite loop of spawning this test and never
+		// actually running the necessary Fatal call.
+		cmd.Env = append(os.Environ(), fmt.Sprintf("%s=%s", testFlag, testFlagVal))
+		cmd.Stderr = &writer
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", testMsg, tc.input))
+		err := cmd.Run()
+		if e, ok := err.(*exec.ExitError); ok && !e.Success() {
+			res, err := writer.Read()
+			require.Nil(t, err, "required successful buffer read")
+			assert.Equal(t, e.ExitCode(), 1, fmt.Sprintf("%s: expected exit code %d, got %d", tc.desc, 1, e.ExitCode()))
+			assert.Equal(t, res, tc.output, fmt.Sprintf("%s: expected output %s got %s", tc.desc, tc.output, res))
```

Comment:
When you assert, do you put first what you expect and then what you got as the params? As in these two lines this seems to be inverse and is a little bit confusing - because it is not consistent.

I will approve, but please change if you see fit.